netket/hilbert/fermions_2nd.py

```diff
@@ -0,0 +1,176 @@
+        Returns:
+            A SpinOrbitalFermions object
+        """
+        spin_size = round(2 * s + 1)
+        spin_states = np.empty(spin_size)
+        for i in range(spin_size):
+            spin_states[i] = 2 * i - round(2 * s)
+        spin_states = spin_states.tolist()
+
+        total_size = n_orbitals * spin_size
+        if n_fermions is None:
+            hilbert = OrbitalFermions(total_size)
+        else:
+            if isinstance(n_fermions, int):
+                hilbert = OrbitalFermions(total_size, n_fermions=n_fermions)
+            else:
+                if not isinstance(n_fermions, Iterable):
+                    raise ValueError("n_fermions must be iterable or int")
+                if len(n_fermions) != spin_size:
+                    raise ValueError(
+                        "list of number of fermions must equal number of spin components"
+                    )
+                spin_hilberts = [OrbitalFermions(n_orbitals, Nf) for Nf in n_fermions]
+                hilbert = TensorHilbert(*spin_hilberts)
+
+        # internally, we store a Fock space or a TensorHilbert of Fock spaces
+        self._fock = hilbert
+        local_states = np.array((0.0, 1.0))
+
+        super().__init__(local_states, N=total_size, constraint_fn=None)
+        self._s = s
+        self.n_fermions = n_fermions
+        self.n_orbitals = n_orbitals
+        self._n_spin_components = spin_size
+        self._spin_states = spin_states
+        # we copy the respective functions, independent of what hilbert space they are
+        self.ptrace = self._fock.ptrace
+        self._numbers_to_states = self._fock._numbers_to_states
+        self._states_to_numbers = self._fock._states_to_numbers
+        self.__mul__ = self._fock.__mul__
+
+    def __repr__(self):
+        _str = "SpinOrbitalFermions(n_orbitals={}".format(self.n_orbitals)
+        if self.n_fermions is not None:
+            _str += ", n_fermions={}".format(self.n_fermions)
+        _str += ")"
+        return _str
+
+    @property
+    def spin(self) -> float:
+        return self._s
+
+    @property
+    def size(self) -> int:
+        return self._fock.size
+
+    @property
+    def _attrs(self):
+        return self._fock._attrs
+
+    @property
+    def is_finite(self) -> bool:
+        return self._fock.is_finite
+
+    @property
+    def n_states(self) -> int:
+        return self._fock.n_states
+
+    @property
+    def n_spin_states(self) -> int:
+        return self._n_spin_components
+
+    @property
+    def spin_states(self) -> List[float]:
+        return tuple(self._spin_states)
+
+    def spin_index(self, sz: float) -> int:
+        return round(2 * sz + 2 * self.spin)
+
+    def get_index(self, site: int, sz: float = None):
+        """go from (site, spin_projection) indices to index in the (tensor) hilbert space"""
+        n_spin_states = self.n_spin_states
+        spin_idx = self.spin_index(sz)
+        return spin_idx * n_spin_states + site
```

Comment:
those two functions seems internal enough that might be better to make them internal (start the name with an underscore).

Still, give them a docstring please.